python/tank/util/shotgun.py

```diff
@@ -224,24 +224,61 @@ def download_url(sg, url, location):
+                                   input `location` value to match the extension
+                                   of the downloaded URL if it can be determined
+                                   and doesn't match already. If set to False, the
+                                   contents of `url` will be downloaded to `location`
+                                   without modification, preserving prior behavior.
+                                   Set this value to True if the file extension for the
+                                   URL to be downloaded cannot be determined in advance,
+                                   for example downloading the content of
+                                   https://my-site.shotgunstudio.com/thumbnail/full/Asset/1227.
+    :returns: Full filepath to the downloaded file. This may have been altered from
+              the input `location` if `use_url_extension` is True and the url's extension
+              was different than `location`.
     :raises: :class:`TankError` on failure.
     """
-    # grab proxy server settings from the shotgun API
-    if sg.config.proxy_handler:
+    # We only need to set the auth cookie for downloads from Shotgun server,
+    # input URLs like: https://my-site.shotgunstudio.com/thumbnail/full/Asset/1227
+    if sg.config.server in url:
+        # this method also handles proxy server settings from the shotgun API
+        __setup_sg_auth_and_proxy(sg)
+    elif sg.config.proxy_handler:
+        # These input URLs have generally already been authenticated and are
+        # in the form: https://sg-media-staging-usor-01.s3.amazonaws.com/9d93f...
+        # %3D&response-content-disposition=filename%3D%22jackpot_icon.png%22.
+        # Grab proxy server settings from the shotgun API
         opener = urllib2.build_opener(sg.config.proxy_handler)
+
         urllib2.install_opener(opener)
     
     # inherit the timeout value from the sg API    
     timeout = sg.config.timeout_secs
     
     # download the given url
     try:
+        request = urllib2.Request(url)
+        if sg.config.server in url:
+            # Currently no public access to Shotgun API _user_agents, so
+            # use directly for now.
+            request.add_header("user-agent", "; ".join(sg._user_agents))
+
         if timeout and sys.version_info >= (2,6):
             # timeout parameter only available in python 2.6+
-            response = urllib2.urlopen(url, timeout=timeout)
+            response = urllib2.urlopen(request, timeout=timeout)
         else:
             # use system default
-            response = urllib2.urlopen(url)
+            response = urllib2.urlopen(request)
+
+        if use_url_extension:
+            # Make sure the disk location has the same extension as the url path.
+            # Would be nice to see this functionality moved to back into Shotgun
+            # API and removed from here.
+            loc_base, loc_ext = os.path.splitext(location)
```

Comment:
rather than trying to replace it, i think it's simpler and more kiss to just append the file extension in case `use_url_extension` has been set to True